CI: Travis: pkg/wireguard/agent/agent_test.go: TestAgent_PeerConfig

Failed test output (https://travis-ci.com/github/cilium/cilium/jobs/501875845) - 

```
=== RUN   Test
coverage: 1.9% of statements in ./...
panic: test timed out after 6m0s
goroutine 13 [running]:
testing.(*M).startAlarm.func1()
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/testing/testing.go:1700 +0xe5
created by time.goFunc
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/time/sleep.go:180 +0x45
goroutine 1 [chan receive, 6 minutes]:
testing.(*T).Run(0xc0003bad80, 0x1c8b969, 0x4, 0x1d64f10, 0x499406)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/testing/testing.go:1239 +0x2da
testing.runTests.func1(0xc0003bac00)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/testing/testing.go:1511 +0x78
testing.tRunner(0xc0003bac00, 0xc00073fda8)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/testing/testing.go:1193 +0xef
testing.runTests(0xc00042c5a0, 0x2bb5d50, 0x1, 0x1, 0xc01ab0b776801048, 0x53d31b01fc, 0x2be5c80, 0x1c990fb)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/testing/testing.go:1509 +0x2fe
testing.(*M).Run(0xc00057b000, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/testing/testing.go:1417 +0x1eb
main.main()
	_testmain.go:1495 +0x1d8
goroutine 19 [chan receive]:
k8s.io/klog/v2.(*loggingT).flushDaemon(0x2be5f20)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/klog/v2/klog.go:1164 +0x8b
created by k8s.io/klog/v2.init.0
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/klog/v2/klog.go:418 +0xdf
goroutine 20 [select, 6 minutes]:
io.(*pipe).Read(0xc0003564e0, 0xc0000b0000, 0x1000, 0x1000, 0x19a6ac0, 0x1, 0xc0000b0000)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/io/pipe.go:57 +0xcb
io.(*PipeReader).Read(0xc0001248d8, 0xc0000b0000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/io/pipe.go:134 +0x4c
bufio.(*Scanner).Scan(0xc00008ff38, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/bufio/scan.go:214 +0xa9
github.com/sirupsen/logrus.(*Entry).writerScanner(0xc0003608c0, 0xc0001248d8, 0xc0003395f0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:59 +0xb7
created by github.com/sirupsen/logrus.(*Entry).WriterLevel
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:51 +0x1b7
goroutine 21 [select, 6 minutes]:
io.(*pipe).Read(0xc000356540, 0xc0000b1000, 0x1000, 0x1000, 0x19a6ac0, 0x1, 0xc0000b1000)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/io/pipe.go:57 +0xcb
io.(*PipeReader).Read(0xc0001248e8, 0xc0000b1000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/io/pipe.go:134 +0x4c
bufio.(*Scanner).Scan(0xc000090f38, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/bufio/scan.go:214 +0xa9
github.com/sirupsen/logrus.(*Entry).writerScanner(0xc0003608c0, 0xc0001248e8, 0xc000339610)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:59 +0xb7
created by github.com/sirupsen/logrus.(*Entry).WriterLevel
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:51 +0x1b7
goroutine 22 [select, 6 minutes]:
io.(*pipe).Read(0xc0003565a0, 0xc0000b2000, 0x1000, 0x1000, 0x19a6ac0, 0x1, 0xc0000b2000)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/io/pipe.go:57 +0xcb
io.(*PipeReader).Read(0xc0001248f8, 0xc0000b2000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/io/pipe.go:134 +0x4c
bufio.(*Scanner).Scan(0xc000091f38, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/bufio/scan.go:214 +0xa9
github.com/sirupsen/logrus.(*Entry).writerScanner(0xc0003608c0, 0xc0001248f8, 0xc000339630)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:59 +0xb7
created by github.com/sirupsen/logrus.(*Entry).WriterLevel
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:51 +0x1b7
goroutine 23 [select, 6 minutes]:
io.(*pipe).Read(0xc000356600, 0xc0000b3000, 0x1000, 0x1000, 0x19a6ac0, 0x1, 0xc0000b3000)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/io/pipe.go:57 +0xcb
io.(*PipeReader).Read(0xc000124908, 0xc0000b3000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/io/pipe.go:134 +0x4c
bufio.(*Scanner).Scan(0xc0000b8f38, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/bufio/scan.go:214 +0xa9
github.com/sirupsen/logrus.(*Entry).writerScanner(0xc0003608c0, 0xc000124908, 0xc000339650)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:59 +0xb7
created by github.com/sirupsen/logrus.(*Entry).WriterLevel
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:51 +0x1b7
goroutine 9 [chan receive, 6 minutes]:
gopkg.in/check%2ev1.(*suiteRunner).runTest(...)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:813
gopkg.in/check%2ev1.(*suiteRunner).run(0xc00057b080, 0x2c15598)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:618 +0x126
gopkg.in/check%2ev1.Run(0x19fca60, 0x2c15598, 0xc000079718, 0x2b047e0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:92 +0x4d
gopkg.in/check%2ev1.RunAll(0xc00008af18, 0x0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:84 +0x96
gopkg.in/check%2ev1.TestingT(0xc0003bad80)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:72 +0x38b
github.com/cilium/cilium/pkg/wireguard/agent.Test(0xc0003bad80)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/wireguard/agent/agent_test.go:39 +0x2b
testing.tRunner(0xc0003bad80, 0x1d64f10)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/testing/testing.go:1193 +0xef
created by testing.(*T).Run
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/testing/testing.go:1238 +0x2b3
goroutine 10 [select, 6 minutes]:
gopkg.in/check%2ev1.(*resultTracker)._loopRoutine(0xc0000e2510)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:464 +0xa5
created by gopkg.in/check%2ev1.(*resultTracker).start
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:444 +0x3f
goroutine 11 [select, 6 minutes]:
golang.org/x/sync/semaphore.(*Weighted).Acquire(0xc000100820, 0x1f072f0, 0xc000120010, 0x1, 0xc000100820, 0xc000574bd0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/golang.org/x/sync/semaphore/semaphore.go:60 +0x345
github.com/cilium/cilium/pkg/lock.(*SemaphoredMutex).RLock(...)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/lock/semaphored_mutex.go:57
github.com/cilium/cilium/pkg/ipcache.(*IPCache).RLock(0xc0001de070)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:131 +0x5d
github.com/cilium/cilium/pkg/wireguard/agent.(*Agent).UpdatePeer(0xc0001de0e0, 0x1c8bcc9, 0x4, 0x1ccef3d, 0x2c, 0xc000055b90, 0x10, 0x10, 0xc000055ba0, 0x10, ...)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/wireguard/agent/agent.go:221 +0x70
github.com/cilium/cilium/pkg/wireguard/agent.(*AgentSuite).TestAgent_PeerConfig(0x2c15598, 0xc000354000)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/wireguard/agent/agent_test.go:141 +0xd5f
reflect.Value.call(0x19fca60, 0x2c15598, 0x213, 0x1c8b6cd, 0x4, 0xc0000bbf08, 0x1, 0x1, 0x2c18320, 0xc000078e40, ...)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/reflect/value.go:476 +0x8e7
reflect.Value.Call(0x19fca60, 0x2c15598, 0x213, 0xc000078f08, 0x1, 0x1, 0x0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/reflect/value.go:337 +0xb9
gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1(0xc000354000)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:775 +0x6b5
gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1(0xc00057b080, 0xc000354000, 0xc0003bd200)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:669 +0x98
created by gopkg.in/check%2ev1.(*suiteRunner).forkCall
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:666 +0x1fe
goroutine 12 [select, 6 minutes]:
golang.org/x/sync/semaphore.(*Weighted).Acquire(0xc000100820, 0x1f072f0, 0xc000120010, 0x40000000, 0x0, 0x0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/golang.org/x/sync/semaphore/semaphore.go:60 +0x345
github.com/cilium/cilium/pkg/lock.(*SemaphoredMutex).Lock(...)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/lock/semaphored_mutex.go:41
github.com/cilium/cilium/pkg/ipcache.(*IPCache).Upsert(0xc0001de070, 0x1c907fc, 0x8, 0xc000055b90, 0x10, 0x10, 0xc000354000, 0x0, 0x3, 0x1c8b2ef, ...)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:259 +0xf1
github.com/cilium/cilium/pkg/wireguard/agent.(*AgentSuite).TestAgent_PeerConfig.func1(0xc0001de070, 0xc000102180)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/wireguard/agent/agent_test.go:130 +0x9e
created by github.com/cilium/cilium/pkg/wireguard/agent.(*AgentSuite).TestAgent_PeerConfig
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/wireguard/agent/agent_test.go:128 +0xcd6
FAIL	github.com/cilium/cilium/pkg/wireguard/agent	360.077s
FAIL
make: *** [Makefile:209: unit-tests] Error 1
```

--- a/pkg/wireguard/agent/agent_test.go
+++ b/pkg/wireguard/agent/agent_test.go
@@ -122,10 +122,31 @@ func (a *AgentSuite) TestAgent_PeerConfig(c *C) {
 	c.Assert(containsIP(k8s1.allowedIPs, pod2IPv4), Equals, true)
 	c.Assert(containsIP(k8s1.allowedIPs, pod2IPv6), Equals, true)
 
-	// Test that IPCache updates are blocked by a concurrent UpdatePeer
-	ipCache.RLock()
-	ipCacheUpdated := make(chan struct{})
+	// Tests that IPCache updates are blocked by a concurrent UpdatePeer.
+	// We test this by issuing an UpdatePeer request while holding
+	// the agent lock (meaning the UpdatePeer call will first take the IPCache
+	// lock and then wait for the agent lock to become available),
+	// then issuing an IPCache update (which will be blocked because
+	// UpdatePeer already holds the IPCache lock), and then releasing the
+	// agent lock to allow both operations to proceed.
+	wgAgent.Lock()
+
+	agentUpdated := make(chan struct{})
+	agentUpdatePending := make(chan struct{})
 	go func() {
+		close(agentUpdatePending)
+		err = wgAgent.UpdatePeer(k8s2NodeName, k8s2PubKey, k8s2NodeIPv4, k8s2NodeIPv6)
+		c.Assert(err, IsNil)
+		close(agentUpdated)
+	}()
+
+	// wait for the above go routine to be scheduled
+	<-agentUpdatePending
+
+	ipCacheUpdated := make(chan struct{})
+	ipCacheUpdatePending := make(chan struct{})
+	go func() {
+		close(ipCacheUpdatePending)
 		// Insert pod3
 		ipCache.Upsert(pod3IPv4Str, k8s2NodeIPv4, 0, nil, ipcache.Identity{ID: 3, Source: source.Kubernetes})
 		ipCache.Upsert(pod3IPv6Str, k8s2NodeIPv6, 0, nil, ipcache.Identity{ID: 3, Source: source.Kubernetes})
@@ -138,10 +159,24 @@ func (a *AgentSuite) TestAgent_PeerConfig(c *C) {
 		close(ipCacheUpdated)
 	}()
 
-	err = wgAgent.UpdatePeer(k8s2NodeName, k8s2PubKey, k8s2NodeIPv4, k8s2NodeIPv6)
-	c.Assert(err, IsNil)
+	// wait for the above go routine to be scheduled
+	<-ipCacheUpdatePending
 
-	ipCache.RUnlock()
+	// At this point we know both go routines have been scheduled. We assume
+	// that they are now both blocked by checking they haven't closed the
+	// channel yet. Thus once release the lock we expect them to make progress
+	select {
+	case <-agentUpdated:
+		c.Fatal("agent update not blocked by agent lock")
+	case <-ipCacheUpdated:
+		c.Fatal("ipcache update not blocked by agent lock")
+	default:
+	}
+
+	wgAgent.Unlock()
+
+	// Ensure that both operations succeeded without a deadlock
+	<-agentUpdated
 	<-ipCacheUpdated
 
 	k8s1 = wgAgent.peerByNodeName[k8s1NodeName]

